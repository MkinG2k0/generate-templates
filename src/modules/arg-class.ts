type TFlag = Record<string, Record<string, string | true | undefined>>

interface Flags {
  template: string[]
  name: string[]
  path: string[]
}
export interface ITemplate {
  templateName: string
  path: string
  flag: Flags
  names: string[]
}

export interface DataArg {
  pathConfig: string
  templates: (ITemplate | undefined)[]
}

interface TArg {
  data: DataArg
}

export class Arg implements TArg {
  data: DataArg = {
    pathConfig: '',
    templates: [],
  }
  constructor(private args: string[]) {
    this.split()
  }

  private split() {
    const [_node, _path, pathConfig, ...options] = this.args

    this.data.pathConfig = pathConfig

    const splitTemplates = options
      .join(' ')
      .split(',')
      .map((value) =>
        value
          .split(' ')
          .map((value) => value.trim())
          .filter((value) => value),
      )

    const data =
      splitTemplates.map((value) => {
        const [template, ...other] = value
        if (!template) return

        const flag: Flags = {
          template: [],
          name: [],
          path: [],
        }

        const names: string[] = []
        let path = ''

        other.map((value) => {
          if (value.includes('--')) {
            const slicedValue = value.slice(2)
            if (names.length === 0) {
              flag.template.push(slicedValue)
            } else if (!path) {
              flag.name.push(slicedValue)
            } else {
              flag.path.push(slicedValue)
            }
          } else if (value.includes('/')) {
            path = value
          } else {
            names.push(value)
          }
        })
        const dataTemplate: ITemplate = { templateName: template, flag, names, path }
        return dataTemplate
      }) || []

    this.data.templates = data
    return data
  }
}
