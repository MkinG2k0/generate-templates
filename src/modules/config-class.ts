import { IConfig } from '../index.js'
import { DataArg } from './arg-class.js'
import { log } from './logs-class.js'
import { Parser } from './parser-class.js'
import { Reader } from './reader-class.js'

interface TConfig {}

export class Config implements TConfig {
  globalConfig: IConfig = {}

  constructor(public args: DataArg, public pathRun: string) {}

  async read() {
    if (!this.args?.pathConfig) {
      log.error('Choose path to config').view(this.args)
      return
    }
    const reader = new Reader(this.args.pathConfig, { path: this.pathRun })
    const data = await reader.read()
    if (data) {
      this.globalConfig = (await Parser.parse(data)) as IConfig
      log.setDebug(log.debug || this.globalConfig.debug)
    }

    return this.globalConfig
  }
}
