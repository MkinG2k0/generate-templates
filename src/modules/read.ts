import nodePath from 'path'
import { TemplateItem } from '../interface/templates-type.js'
import { ITemplate } from '../modules/arg-class.js'
import { Config } from './config-class.js'
import { log } from './logs-class.js'
import { Reader } from './reader-class.js'

export class ReadTemplates {
  constructor(public config: Config) {}

  // Существует ли путь до темлейтов который указан в конфиге
  async findTemplate() {
    // Достаем из аргументов название темлейтов ([page,comp])
    if (!this.config.args) return

    const find = this.config.args.templates.map(async (template) => {
      if (!template) {
        log.error(`Can not find template pls check console`).view(template)
        return
      }

      const { templateName } = template

      // Достаем по имени темлейта объект из конфига
      const localConfig = this.config.globalConfig?.templates?.[templateName]

      // Если нашли объект то значит что темлейт найден
      if (!localConfig) {
        // Если не нашли темлейт, то бросаем ошибку
        log.error(`Can not find config ${templateName} pls check template file`).view(template)
        return
      }

      // Инициализируем темлейт
      return await this.initTemplate(localConfig, template)
    })

    return Promise.all(find)
  }

  private async initTemplate(localConfig: TemplateItem, template: ITemplate) {
    // Путь до файлов в теплейте
    const pathToTemplate = nodePath.join(this.config.pathRun, localConfig.template)
    // Существует ли файл
    const isExists = await Reader.isExists(pathToTemplate)

    if (isExists) {
      return { localConfig, template }
    } else {
      // В противном случае бросаем ошибку
      log.error(`Can not find file to "${pathToTemplate}"`)
    }
  }
}
