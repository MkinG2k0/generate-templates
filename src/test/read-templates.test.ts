import { describe, expect } from 'vitest'
import { Arg } from '../modules/arg-class.js'
import { Config } from '../modules/config-class.js'
import { ReadTemplates } from '../modules/read.js'
import { processMock } from './mock/process.js'

describe('ReadTemplates class', async () => {
  const arg = new Arg(processMock.argv)
  const path = processMock.cwd()
  const config = new Config(arg.data, path)
  const dataConfig = await config.read()

  test('ReadTemplates args', async () => {
    const readTemplates = new ReadTemplates(config)

    expect(dataConfig).toBeDefined()

    const dataTemplate = await readTemplates.findTemplate()

    expect(dataTemplate).matchSnapshot()
  })
})
