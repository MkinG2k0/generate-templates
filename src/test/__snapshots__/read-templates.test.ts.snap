// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ReadTemplates class > ReadTemplates args 1`] = `
[
  {
    "localConfig": {
      "generate": "generated/Pages/",
      "replaceNameFolder": false,
      "template": "templates/page",
    },
    "template": {
      "flag": {
        "name": [],
        "path": [],
        "template": [],
      },
      "names": [
        "Main",
        "Auth",
      ],
      "path": "",
      "templateName": "page",
    },
  },
]
`;
